### DISCUSSION TITLE
Thermostat->targetTemperature->getVal() float / int problem

### DISCUSSION THREAD

**Discussion by TomBBPL** [HomeSpan/HomeSpan#184] [Q&A] [answered]
Hi, Thank you for the awesome library!
so,
I'm trying to make a thermostat.
Everything works almost fine, but I can't get the floating point temperature. 
On the iPhone screen displays correctly, I can change the values in the 0.5 step, 
but a "targetTemperature" variable that is always given an integer value
What am I doing wrong, how to make the variable "targetTemperature" have an actual floating point value ?

struct DEV_iSence:Service::Thermostat{
  SpanCharacteristic *currentHeatingCoolingState;
  SpanCharacteristic *targetHeatingCoolingState;
  SpanCharacteristic *currentTemperature;
  SpanCharacteristic *targetTemperature;
  SpanCharacteristic *temperatureDisplayUnits;

    DEV_iSence() : Service::Thermostat(){
      currentHeatingCoolingState    = new Characteristic::CurrentHeatingCoolingState(1);
      targetHeatingCoolingState = new Characteristic::TargetHeatingCoolingState(1);
      currentTemperature   = new Characteristic::CurrentTemperature(19.5); //TP_IN
      currentTemperature->setRange(0, 50, 0.5);
      targetTemperature = new Characteristic::TargetTemperature(21.5); //TP_SetPoint
      targetTemperature->setRange(0, 50, 0.5); 
      temperatureDisplayUnits = new Characteristic::TemperatureDisplayUnits(0);
    }
...
void loop(){
...
Serial.println(targetTemperature->getVal());  <- it always print int.
...
}

**Comment by SnowballsChance** [comment, marked as answer]
I had this problem last night too.

Try this

`targetTemperature->getVal<float>()`


This forces it to return a float.

I found it in the[ service loops example](https://github.com/HomeSpan/HomeSpan/blob/master/examples/12-ServiceLoops/DEV_Sensors.h).

Hope this helps

**Reply by TomBBPL** [reply]
Works great, thanks a lot :)
